### PR TITLE
Refactor out uses of innerRepo

### LINKED ISSUE #18342: Data race in RepositoryResolver
Failed CI run here: https://buildkite.com/sourcegraph/sourcegraph/builds/87228#f11e1202-ca31-4c31-a7d5-d9f8045ff3bc

<details>

```
WARNING: DATA RACE
Write at 0x00c000600090 by goroutine 54:
  github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend.(*RepositoryResolver).hydrate.func1()
      /buildkite/builds/buildkite-agent-5f65f55c76-6tvdh-1/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend/repository.go:320 +0x31c
  sync.(*Once).doSlow()
      /root/.asdf/installs/golang/1.15/go/src/sync/once.go:66 +0x109
  sync.(*Once).Do()
      /root/.asdf/installs/golang/1.15/go/src/sync/once.go:57 +0x68
  github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend.(*RepositoryResolver).hydrate()
      /buildkite/builds/buildkite-agent-5f65f55c76-6tvdh-1/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend/repository.go:308 +0x7a
  github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend.(*RepositoryResolver).repo()
      /buildkite/builds/buildkite-agent-5f65f55c76-6tvdh-1/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend/repository.go:100 +0x70
  github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend.(*RepositoryResolver).Commit()
      /buildkite/builds/buildkite-agent-5f65f55c76-6tvdh-1/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend/repository.go:158 +0x45
  github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend.NewPreviewRepositoryComparisonResolver()
      /buildkite/builds/buildkite-agent-5f65f55c76-6tvdh-1/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend/preview_repository_comparison.go:20 +0xe5
  github.com/sourcegraph/sourcegraph/enterprise/internal/campaigns/resolvers.(*changesetResolver).Diff()
      /buildkite/builds/buildkite-agent-5f65f55c76-6tvdh-1/sourcegraph/sourcegraph/enterprise/internal/campaigns/resolvers/changeset.go:466 +0xc24
  runtime.call64()
      /root/.asdf/installs/golang/1.15/go/src/runtime/asm_amd64.s:541 +0x3d
  reflect.Value.Call()
      /root/.asdf/installs/golang/1.15/go/src/reflect/value.go:336 +0xd8
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection.func2()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:203 +0x8fb
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:223 +0x3a4
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections.func1()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:83 +0x284

Previous read at 0x00c000600090 by goroutine 106:
  github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend.(*RepositoryResolver).Name()
      /buildkite/builds/buildkite-agent-5f65f55c76-6tvdh-1/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend/repository.go:105 +0x44
  runtime.call32()
      /root/.asdf/installs/golang/1.15/go/src/runtime/asm_amd64.s:540 +0x3d
  reflect.Value.Call()
      /root/.asdf/installs/golang/1.15/go/src/reflect/value.go:336 +0xd8
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection.func2()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:203 +0x8fb
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:223 +0x3a4
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:90 +0x1c4
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelectionSet()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:258 +0x444
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:237 +0x5d5
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections.func1()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:83 +0x284

Goroutine 54 (running) created at:
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:79 +0x8e4
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelectionSet()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:258 +0x444
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:237 +0x5d5
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections.func1()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:83 +0x284

Goroutine 106 (finished) created at:
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:79 +0x8e4
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelectionSet()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:258 +0x444
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:237 +0x5d5
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections.func1()
      /root/.asdf/installs/golang/1.15/packages/pkg/mod/github.com/sourcegraph/graphql-go@v0.0.0-20201007040903-ec61a5417d66/internal/exec/exec.go:83 +0x284
```

</details>

### PR DESCRIPTION
**Pull request by camdencheek**
This turned into a _much_ larger change than I was hoping for, especially since there is a chance this will become obsolete if I ever get through the `graphqlbackend` refactor, but I think it's necessary in order to get rid of all race conditions with hydration. 

Essentially, the issue was that there were _many_ parts of the codebase that were accessing `innerRepo` directly, which is problematic if we want the methods of `RepositoryResolver` to be safe to use from multiple goroutines. 

Summary of the changes:
- When creating a `RepositoryResolver`, always save the name and ID outside of the hydrated repo. This way, they can always be safely accessed even if a concurrent hydration is occuring
- Always create a `RepositoryResolver` with `NewRepositoryResolver` so that the name and ID are saved
- Convert all accesses of `RepositoryResolver.innerRepo.Name` to `RepositoryResolver.name`

Fixes #18342 

I added Campaigns as a reviewer because the last commit here touches campaigns-specific code (there was a slice that was being initialized incorrectly)

### CODE REVIEW

**Comment by camdencheek** [comment]
Looks like I broke a resolver. Fix incoming

**Comment by camdencheek** [comment]
Tests still failing seemingly due to a possibly unrelated issue.

It appears that `reposByID[c.RepoID]` is `nil` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@b93604623518210c04a6e02df7085d051dddb3be/-/blob/enterprise/internal/campaigns/resolvers/changeset_connection.go#L59), which is causing a nil pointer error when trying to access name further down the line. I'm also getting some failures about repo permissions that could be related to [this](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@b93604623518210c04a6e02df7085d051dddb3be/-/blob/enterprise/internal/campaigns/resolvers/changeset_connection.go#L101-L102). 

The initial race condition I was seeing was because `RepositoryResolver.Name()` and `RepositoryResolver.Commit()`, were being called concurrently, and `Commit()` was mutating `RepositoryResolver.innerRepo`. This PR should fix that, but it's unclear to me whether I caused these new failures or whether they were due to a confounding change. 

cc @sourcegraph/campaigns

File: cmd/frontend/graphqlbackend/search_repositories.go [hunk 1]
```diff
@@ -116,6 +116,9 @@ func repoRevsToSearchResultResolver(ctx context.Context, repos []*search.Reposit
 			revs = r.RevSpecs()
 		}
 		for _, rev := range revs {
+			rr := NewRepositoryResolver(r.Repo.ToRepo())
+			rr.icon = repoIcon
+			rr.rev = rev
 			results = append(results, &RepositoryResolver{innerRepo: r.Repo.ToRepo(), icon: repoIcon, rev: rev})
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_repositories.go:122]: not using rr here. I'd recommend renaming innerRepo without the use of an IDE to ensure you have caught all old uses of it (ie if it builds, you know you have done it)

**Review by keegancsmith — APPROVED** [review]

File: internal/campaigns/changeset.go [hunk 2]
```diff
@@ -730,7 +730,7 @@ func (cs Changesets) RepoIDs() []api.RepoID {
 	for _, c := range cs {
 		repoIDMap[c.RepoID] = struct{}{}
 	}
-	repoIDs := make([]api.RepoID, len(repoIDMap))
+	repoIDs := make([]api.RepoID, 0, len(repoIDMap))
```
Comment (eseliger) [inline internal/campaigns/changeset.go:733]: Oo

**Review by eseliger — APPROVED** [review]

**Comment by mrnugget** [comment]
Good stuff, thanks @camdencheek!

**Pull request merged**